src/store.rs

```diff
@@ -907,68 +615,89 @@ impl<F: LurkField> Store<F> {
-        ));
-
-        if ptr.raw.is_opaque() {
-            let is_keyword = ptr.tag == ExprTag::Key;
+    pub fn fetch_symcons(&self, ptr: &Ptr<F>) -> Option<(Ptr<F>, Ptr<F>)> {
+        match (ptr.tag, ptr.raw) {
+            (ExprTag::Sym, RawPtr::Null) => None,
+            (ExprTag::Sym, RawPtr::Index(x)) => {
+                let (car, cdr) = self.sym_store.get_index(x)?;
+                Some((*car, *cdr))
+            }
+            _ => None,
+        }
+    }
 
-            return Some(Sym::new_opaque(is_keyword));
+    pub fn fetch_sym(&self, ptr: &Ptr<F>) -> Option<Symbol> {
+        let mut ptr = *ptr;
+        let mut path = Vec::new();
+        while let Some((car, cdr)) = self.fetch_symcons(&ptr) {
+            let string = self.fetch_string(&car)?;
+            path.push(string);
+            ptr = cdr
         }
+        Some(Symbol::Sym(path.into_iter().rev().collect()))
+    }
 
+    pub fn fetch_key(&self, ptr: &Ptr<F>) -> Option<Symbol> {
+        let symbol = self.fetch_sym(&Ptr {
+            tag: ExprTag::Sym,
+            raw: ptr.raw,
+            _f: ptr._f,
+        })?;
+        Some(Symbol::Key(symbol.path()))
+    }
+
+    pub fn fetch_symbol(&self, ptr: &Ptr<F>) -> Option<Symbol> {
         if ptr.tag == ExprTag::Nil {
-            return Some(Sym::new(".LURK.NIL".into()));
-        };
-        self.sym_store
-            .0
-            .resolve(SymbolUsize::try_from_usize(ptr.raw.idx()?).unwrap())
-            .map(|s| match ptr.tag {
-                ExprTag::Sym => Sym::new_sym(s.into()),
-                ExprTag::Key => Sym::new_key(s.into()),
-                _ => unreachable!(),
-            })
+            Some(Symbol::nil())
+        } else if ptr.tag == ExprTag::Key {
+            self.fetch_key(ptr)
+        } else {
+            self.fetch_sym(ptr)
+        }
+    }
+
+    pub fn fetch_strcons(&self, ptr: &Ptr<F>) -> Option<(Ptr<F>, Ptr<F>)> {
+        match (ptr.tag, ptr.raw) {
+            (ExprTag::Str, RawPtr::Null) => None,
+            (ExprTag::Str, RawPtr::Index(x)) => {
+                let (car, cdr) = self.str_store.get_index(x)?;
+                Some((*car, *cdr))
+            }
+            _ => None,
+        }
     }
 
-    pub fn fetch_str(&self, ptr: &Ptr<F>) -> Option<&str> {
-        debug_assert!(matches!(ptr.tag, ExprTag::Str));
-        let symbol = SymbolUsize::try_from_usize(ptr.raw.idx()?)?;
-        self.str_store.0.resolve(symbol)
+    pub fn fetch_string(&self, ptr: &Ptr<F>) -> Option<String> {
+        let mut string = String::new();
+        if *ptr == self.strnil() {
+            return Some(string);
+        }
+        let mut strcons = self.fetch_strcons(ptr);
+        if let Some(..) = strcons {
+            while let Some((car, cdr)) = strcons {
+                let chr = self.fetch_char(&car)?;
+                string.push(chr);
+                strcons = self.fetch_strcons(&cdr)
+            }
+            return Some(string);
+        }
+        None
     }
```

Comment:
This was hard to follow. It would arguably be clearer if you just checked whether `string` is empty (and if so, return `None`) before returning on line 698.

Reply:
Agreed that this implementation was hard to follow, have rewritten it to do the loop directly:

```rust
    pub fn fetch_string(&self, ptr: &Ptr<F>) -> Option<String> {
        let mut string = String::new();
        let mut ptr = ptr;
        loop {
            match (ptr.tag, ptr.raw) {
                (ExprTag::Str, RawPtr::Null) => return Some(string),
                (ExprTag::Str, RawPtr::Index(x)) => {
                    let (car, cdr) = self.str_store.get_index(x)?;
                    let chr = self.fetch_char(&car)?;
                    string.push(chr);
                    ptr = cdr
                }
                _ => return None,
            }
        }
    }
```

which I think is much cleaner (and actually fixes a subtle correctness issue the prior implementation had)